miner/collator.go

```diff
@@ -0,0 +1,245 @@
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// BlockState represents a block-to-be-mined, which is being assembled. A collator
+// can add transactions, and finish by calling Commit.
+type BlockState interface {
+	// AddTransactions adds the sequence of transactions to the blockstate. Either all
+	// transactions are added, or none of them. In the latter case, the error
+	// describes the reason why the txs could not be included.
+	AddTransactions(sequence types.Transactions) error
+	// Commit signals that the collation is finished, and the block is ready
+	// to be sealed. After calling Commit, no more transactions may be added.
+	Commit()
+
+	Gas() (remaining uint64)
+	Coinbase() common.Address
+	BaseFee() *big.Int
+	Signer() types.Signer
+}
+
+// Collator is something that can assemble a block.
+type Collator interface {
+	CollateBlock(bs BlockState, txs types.Transactions, interrupt *int32, isSealing bool) error
+}
+
+// Pool is an interface to the transaction pool
+type Pool interface {
+	Pending(bool) (map[common.Address]types.Transactions, error)
+	Locals() []common.Address
+}
+
+// blockState is an implementation of BlockState
+type blockState struct {
+	state    *state.StateDB
+	logs     []*types.Log
+	worker   *worker
+	coinbase common.Address
+	baseFee  *big.Int
+	signer   types.Signer
+}
+
+// Coinbase returns the miner-address of the block being mined
+func (bs *blockState) Coinbase() common.Address {
+	return bs.coinbase
+}
+
+// Basefee returns the basefee for the current block
+func (bs *blockState) BaseFee() *big.Int {
+	return bs.baseFee
+}
+
+// Signer returns the block-specific signer
+func (bs *blockState) Signer() types.Signer {
+	return bs.signer
+}
+
+func (bs *blockState) Commit() {
+	w := bs.worker
+	// We don't push the pendingLogsEvent while we are mining. The reason is that
+	// when we are mining, the worker will regenerate a mining block every 3 seconds.
+	// In order to avoid pushing the repeated pendingLog, we disable the pending log pushing.
+	if !w.isRunning() && len(bs.logs) > 0 {
+
+		// make a copy, the state caches the logs and these logs get "upgraded" from pending to mined
+		// logs by filling in the block hash when the block was mined by the local miner. This can
+		// cause a race condition if a log was "upgraded" before the PendingLogsEvent is processed.
+		cpy := make([]*types.Log, len(bs.logs))
+		for i, l := range bs.logs {
+			cpy[i] = new(types.Log)
+			*cpy[i] = *l
+		}
+		w.pendingLogsFeed.Send(cpy)
+	}
+}
+
+func (bs *blockState) AddTransactions(sequence types.Transactions) error {
+	var (
+		w      = bs.worker
+		snap   = w.current.state.Snapshot()
+		err    error
+		logs   []*types.Log
+		tcount = w.current.tcount
+	)
+	for _, tx := range sequence {
+		if w.current.gasPool.Gas() < params.TxGas {
+			log.Trace("Not enough gas for further transactions", "have", w.current.gasPool, "want", params.TxGas)
+			return core.ErrGasLimitReached
+		}
```

Comment:
This needs to revert any txs in the sequence that have already been added. You can just set `err = core.ErrGasLimitReached` and `break` out of the loop, instead of returning